// @mintlify/scraping/src/scraping/detectFramework.ts
import * as cheerio from 'cheerio';

export const frameworks = ['docusaurus', 'gitbook', 'readme', 'intercom'] as const;
export type Framework = (typeof frameworks)[number];

export type FrameworkHint =
  | {
      framework: 'docusaurus';
      version: '1' | '2' | '3';
    }
  | {
      framework: 'gitbook' | 'readme' | 'intercom' | undefined;
    };

export function detectFramework(html: string): FrameworkHint {
  const $ = cheerio.load(html);
  const docusaurusMeta = $('meta[name="generator"]');

  if (
    docusaurusMeta.length > 0 &&
    docusaurusMeta.attr('content') &&
    typeof docusaurusMeta.attr('content') === 'string' &&
    (docusaurusMeta.attr('content') as string).includes('Docusaurus')
  ) {
    const metaAttrString = docusaurusMeta.attr('content') as string;
    if (metaAttrString.includes('v3')) {
      return { framework: 'docusaurus', version: '3' };
    }
    if (metaAttrString.includes('v2')) {
      return { framework: 'docusaurus', version: '2' };
    } else if (metaAttrString.includes('v1')) {
      console.warn(
        'WARNING: We detected Docusaurus version 1 but we only support scraping versions 2 and 3.'
      );
      return { framework: 'docusaurus', version: '1' };
    }
  }

  const isGitBook = $('.gitbook-root').length > 0;
  if (isGitBook) {
    return { framework: 'gitbook' };
  }

  const isReadMe = $('meta[name="readme-deploy"]').length > 0;
  if (isReadMe) {
    return { framework: 'readme' };
  }

  const isIntercom = $("meta[name='intercom:trackingEvent']").length > 0;
  if (isIntercom) {
    return { framework: 'intercom' };
  }

  return { framework: undefined };
}
