// @mintlify/validation/dist/mint-config/schemas/apiReference.js
import { apiParamFieldOptions, requestExampleLanguages } from '@mintlify/models';
import { z } from 'zod';
export const openApiSchema = z
    .union([z.string(), z.array(z.string())])
    .describe('A string or an array of strings of absolute or relative urls pointing to your OpenAPI files');
export const apiSchema = z
    .object({
    baseUrl: z.union([z.string().url(), z.array(z.string().url())]).optional(),
    auth: z
        .object({
        method: z.enum(['bearer', 'basic', 'key', 'cobo']).optional(),
        name: z.string().optional(),
        inputPrefix: z.string().optional(),
    })
        .strict()
        .optional(),
    playground: z
        .object({
        mode: z.enum(['show', 'simple', 'hide']).optional().default('show'),
        disableProxy: z.boolean().optional(),
    })
        .optional(),
    request: z
        .object({
        example: z
            .object({
            showOptionalParams: z.boolean().optional(),
            languages: z.array(z.union([z.enum(requestExampleLanguages), z.string()])).optional(),
        })
            .optional(),
    })
        .optional(),
    maintainOrder: z.boolean().optional(),
    paramFields: z
        .object({
        expanded: z.enum(apiParamFieldOptions).optional(),
    })
        .optional(),
})
    .strict();
