// @mintlify/scraping/bin/scraping/scrapeSectionCommands.js
import axios from 'axios';
import { startBrowser } from '../browser.js';
import { getOrigin } from '../util.js';
import { detectFramework } from './detectFramework.js';
import { scrapeSection } from './scrapeSection.js';
import { scrapeIntercomSection } from './site-scrapers/Intercom/scrapeIntercomSection.js';
import openNestedDocusaurusMenus from './site-scrapers/openNestedDocusaurusMenus.js';
import openNestedGitbookMenus from './site-scrapers/openNestedGitbookMenus.js';
import { scrapeDocusaurusSection } from './site-scrapers/scrapeDocusaurusSection.js';
import { scrapeGitBookSection } from './site-scrapers/scrapeGitBookSection.js';
import { scrapeReadMeSection } from './site-scrapers/scrapeReadMeSection.js';
export async function scrapeSectionAxiosWrapper(url, overwrite, scrapeFunc) {
    const res = await axios.get(url);
    const html = res.data;
    await scrapeSection(scrapeFunc, html, getOrigin(url), overwrite, undefined);
    process.exit(0);
}
export async function scrapeDocusaurusSectionCommand(url, overwrite, version // "1" | "2" | "3"
) {
    await scrapeSectionOpeningAllNested(url, overwrite, openNestedDocusaurusMenus, scrapeDocusaurusSection, version);
}
export async function scrapeGitbookSectionCommand(url, overwrite) {
    await scrapeSectionOpeningAllNested(url, overwrite, openNestedGitbookMenus, scrapeGitBookSection);
}
async function scrapeSectionOpeningAllNested(url, overwrite, openLinks, scrapeFunc, version) {
    const browser = await startBrowser();
    const page = await browser.newPage();
    await page.goto(url, {
        waitUntil: 'networkidle2',
    });
    const html = await openLinks(page);
    void browser.close();
    await scrapeSection(scrapeFunc, html, getOrigin(url), overwrite, version);
    process.exit(0);
}
export async function scrapeSectionAutomatically(url, overwrite, frameworkHint) {
    const res = await axios.get(url);
    const html = res.data;
    frameworkHint = frameworkHint.framework ? frameworkHint : detectFramework(html);
    validateFramework(frameworkHint.framework);
    console.log('Detected framework: ' + frameworkHint.framework);
    switch (frameworkHint.framework) {
        case 'docusaurus':
            await scrapeDocusaurusSectionCommand(url, overwrite, frameworkHint.version);
            break;
        case 'gitbook':
            await scrapeGitbookSectionCommand(url, overwrite);
            break;
        case 'readme':
            await scrapeSectionAxiosWrapper(url, overwrite, scrapeReadMeSection);
            break;
        case 'intercom':
            await scrapeSectionAxiosWrapper(url, overwrite, scrapeIntercomSection);
            break;
    }
}
function validateFramework(framework) {
    if (!framework) {
        console.log('Could not detect the framework automatically. We only support Docusaurus (V2 and V3), GitBook, and ReadMe.');
        process.exit();
    }
}
