// @mintlify/prebuild/dist/prebuild/update/read/getOpenApiFilesFromMintConfig.js
import SwaggerParser from '@apidevtools/swagger-parser';
import axios from 'axios';
import yaml from 'js-yaml';
export const getOpenApiFilesFromMintConfig = async (mintConfig) => {
    const openApiFiles = [];
    // Download OpenApi file if url is provided
    if (mintConfig.openapi &&
        typeof mintConfig.openapi === 'string' &&
        mintConfig.openapi.startsWith('https')) {
        const specFromUrl = await getOpenApiDocumentFromUrl(mintConfig.openapi);
        openApiFiles.push({
            filename: 'openapi-from-url',
            spec: specFromUrl,
        });
    }
    return openApiFiles;
};
export const getOpenApiDocumentFromUrl = async (url) => {
    const { data } = await axios.get(url, {
        responseType: 'text',
        transformResponse: (res) => {
            // Disable automatic JSON parsing
            return res;
        },
    });
    let openApiDocument;
    if (url.endsWith('yaml') || url.endsWith('yml')) {
        openApiDocument = yaml.load(data);
    }
    else {
        openApiDocument = JSON.parse(data);
    }
    if (openApiDocument == undefined) {
        throw Error('Could not parse OpenAPI document.');
    }
    // Before uploading, just check to see if spec is valid - don't dereference.
    const validatedApi = await SwaggerParser.validate(structuredClone(openApiDocument));
    if (!('openapi' in validatedApi)) {
        throw Error('OpenAPI v2 not supported');
    }
    return openApiDocument;
};
